runtime/src/iree/hal/utils/libmpi_test.cc

```diff
@@ -88,4 +89,188 @@ TEST_F(LibmpiTest, MPIErrorToIREEStatus) {
               StatusIs(StatusCode::kInternal));
 }
 
-}  // namespace
+TEST_F(LibmpiTest, Bcast) {
+  const int N = 100;
+  char array[N];
+  int root = 1;
+
+  ASSERT_GT(world_size, 1)
+      << "This test requires mpirun with more than one node";
+
+  if (rank == root) memset(array, 'r', N);
+
+  IREE_EXPECT_OK(
+      MPI_RESULT_TO_STATUS(&symbols,
+                           MPI_Bcast(array, N, IREE_MPI_BYTE(&symbols), root,
+                                     IREE_MPI_COMM_WORLD(&symbols)),
+                           "MPI_Bcast"));
+
+  for (int i = 0; i < N; i++)
+    EXPECT_EQ(array[i], 'r') << "[" << rank << "] failed at index " << i;
+}
+
+TEST_F(LibmpiTest, Gather) {
+  const int N = 100;
+  char sendarray[N];
+  int root = 1;
+  char *recvbuf = NULL;
+
+  ASSERT_GT(world_size, 1)
+      << "This test requires mpirun with more than one node";
+
+  memset(sendarray, (char)(rank + 1), N);
+
+  if (rank == root) {
+    recvbuf = (char *)malloc(world_size * N * sizeof(char));
+    ASSERT_NE(recvbuf, nullptr);
+  }
+  IREE_EXPECT_OK(MPI_RESULT_TO_STATUS(
+      &symbols,
+      MPI_Gather(sendarray, N, IREE_MPI_BYTE(&symbols), recvbuf, N,
+                 IREE_MPI_BYTE(&symbols), root, IREE_MPI_COMM_WORLD(&symbols)),
+      "MPI_Gather"));
+
+  if (rank == root) {
+    for (int i = 0; i < N * world_size; i++) {
+      EXPECT_EQ(recvbuf[i], (char)((i / N) + 1))
+          << "[" << rank << "] failed at index " << i;
+    }
+    free(recvbuf);
+  }
+}
+
+TEST_F(LibmpiTest, Scatter) {
+  const int N = 100;
+  char recvbuf[N];
```

Comment:
zero initialize or initialize with a sentinel that is deterministic (if one of these fails you need to ensure it fails consistently)
```suggestion
  char recvbuf[N] = {0};
```